pkg/controller/controller.go

```diff
@@ -374,6 +376,66 @@ func (c *CSIControllerService) ListSnapshots(context.Context, *csi.ListSnapshots
 }
 
 // ControllerExpandVolume is not implemented yet
-func (c *CSIControllerService) ControllerExpandVolume(context.Context, *csi.ControllerExpandVolumeRequest) (*csi.ControllerExpandVolumeResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "not implemented yet")
+func (c *CSIControllerService) ControllerExpandVolume(ctx context.Context, req *csi.ControllerExpandVolumeRequest) (*csi.ControllerExpandVolumeResponse, error) {
+	ll := c.log.WithFields(logrus.Fields{
+		"method":   "ControllerExpandVolume",
+		"volumeID": req.GetVolumeId(),
+	})
+	ll.Infof("Processing request: %v", req)
+	var (
+		volID     = req.GetVolumeId()
+		lvg       = &lvgcrd.LVG{}
+		ctxWithID = context.WithValue(context.Background(), base.RequestUUID, volID)
+	)
+
+	if volID == "" {
+		return nil, status.Error(codes.InvalidArgument, "Volume name missing in request")
+	}
+	if req.GetVolumeCapability() == nil {
+		return nil, status.Error(codes.InvalidArgument, "Volume capabilities missing in request")
+	}
+
+	volume := c.crHelper.GetVolumeByID(volID)
+
+	if volume == nil {
+		return nil, status.Error(codes.NotFound, "Volume doesn't exist")
+	}
+
+	if !util.IsStorageClassLVG(volume.Spec.StorageClass) {
+		return nil, status.Error(codes.FailedPrecondition, "Volume doesn't have LVG storage class")
+	}
+
+	if volume.Spec.Size > req.GetCapacityRange().GetRequiredBytes() {
+		return nil, status.Error(codes.OutOfRange, "Required bytes are less than existing volume size")
+	}
+
+	if volume.Spec.Size == req.GetCapacityRange().GetRequiredBytes() {
+		return &csi.ControllerExpandVolumeResponse{
+			CapacityBytes:         0,
+			NodeExpansionRequired: false,
+		}, nil
+	}
+
+	if err := c.k8sclient.ReadCR(ctxWithID, volume.Spec.Location, "", lvg); err != nil {
+		ll.Errorf("Failed to get LVG by location %s", volume.Spec.Location)
+		return nil, status.Error(codes.Internal, "Unable to read LVG")
+	}
+	if lvg.Spec.Size < req.GetCapacityRange().GetRequiredBytes() {
```

Comment:
Could you please describe why you are using lvg size here? Maybe it will be more correct to use LVG AC size here? Additionally, we need to decrease LVG AC size here.